gwcs/wcs.py

```diff
@@ -1606,31 +1680,119 @@ def to_fits_sip(self, bounding_box=None, max_pix_error=0.25, degree=None,
-                hdr['ap_order'] = ipdegree
-                hdr['bp_order'] = ipdegree
-        else:
-            hdr['ctype1'] = 'RA---TAN'
-            hdr['ctype2'] = 'DEC--TAN'
+                hdr['sipiverr'] = (max_inv_resid, 'Max diff for inverse (pixels)')
+
+        # CD matrix
+        # We do not update comments to typical comments for CD matrix elements
+        # (such as 'partial of second axis coordinate w.r.t. y'), because
+        # when input frame has number of axes > 2, then imaging
+        # axes arbitrary.
+        del hdr['CDELT?']
+
+        old_nlon = 1 if nlon < nlat else 2
+        old_nlat = 2 if nlon < nlat else 1
+        hdr.rename_keyword(f'PC{old_nlon}_1', f'CD{nlon}_{iax1}')
+        hdr.rename_keyword(f'PC{old_nlon}_2', f'CD{nlon}_{iax2}')
+        hdr.rename_keyword(f'PC{old_nlat}_1', f'CD{nlat}_{iax1}')
+        hdr.rename_keyword(f'PC{old_nlat}_2', f'CD{nlat}_{iax2}')
+
+        # remap axes:
+        axis_rename = {}
+        if iax1 != 1:
+            axis_rename['CRPIX1'] = f'CRPIX{iax1}'
+        if iax2 != 2:
+            axis_rename['CRPIX2'] = f'CRPIX{iax2}'
+        for kwd in ['CRVAL', 'CTYPE', 'CUNIT']:
+            for iold, inew in [(1, nlon), (2, nlat)]:
+                if iold != inew:
+                    axis_rename[f'{kwd:s}{iold:d}'] = f'{kwd:s}{inew:d}'
+
+        old_cards = hdr.cards
+        new_cards = []
+        for c in old_cards:
+            if c[0] in axis_rename:
+                c = fits.Card(keyword=axis_rename[c.keyword], value=c.value, comment=c.comment)
+            new_cards.append(c)
+
+        hdr = fits.Header(new_cards)
+        hdr['WCSAXES'] = naxis
+        hdr.insert('WCSAXES', ('WCSNAME', f"'{self.output_frame.name}' gWCS converted to FITS WCS"))
 
-        hdr['cd1_1'] = cdmat[0][0]
-        hdr['cd1_2'] = cdmat[0][1]
-        hdr['cd2_1'] = cdmat[1][0]
-        hdr['cd2_2'] = cdmat[1][1]
         return hdr
 
+    def _separable_axes_sets(self):
+        corr_mat = self.axis_correlation_matrix
+        sets = [set(np.flatnonzero(r)) for r in corr_mat.T]
+
+        k = 0;
+        while len(sets) - 1 > k:
+            for m in range(len(sets) - 1, k, -1):
+                if sets[k].isdisjoint(sets[m]):
+                    continue
+                sets[k] = sets[k].union(sets[m])
+                del sets[m]
+            k += 1
+
+        mapping = {k: tuple(np.flatnonzero(r)) for k, r in enumerate(corr_mat)}
+
+        return sets, mapping
+
+    def _separable_frames_and_axes_groups(self):
+        """ Return a list of tuples of (coordinate frames, list of tuples
+        of separable output axes) and mapping of output axes to input axes.
+        """
+        axes_sets, mapping = self._separable_axes_sets()
+        frame_axes_groups = []
+
+        if isinstance(self.output_frame, cf.CompositeFrame):
+            frames = self.output_frame.frames
+        else:
+            frames = [self.output_frame]
+
+        for frame in frames:
```

Comment:
Again, this could use more comments about what each block is doing with regard to coordinate frames and indices.